### ISSUE TITLE
missing braces and typo in duplicated beginLoggingTo warning string

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7916]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| d.vinella reported|
|-|-|
|Trac ID|trac#7916|
|Type|defect|
|Created|2015-05-27 12:05:05Z|
|Branch|https://github.com/twisted/twisted/tree/beginLoggingTo-7916-2|

In twisted/logger/_global:27, MORE_THAN_ONCE_WARNING
"<{fileThen:logThen}>" should be "<{fileThen}:{lineThen}>"

Attachments:

* [logger_global.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/24577033564ece8ba2708d11ce3f4d0e5407c2f9.patch) (540 bytes) - added by d.vinella on 2015-05-27 12:05:34Z - 
* [7916-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/c3910d0c4c8c06e22a9db4bc213eeafb27ce1d05.patch) (540 bytes) - added by d.vinella on 2015-05-28 20:07:24Z - patch with correct name, as per guidelines
* [7916-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/84776d712265522dd036f18fd8363945691a0b73.patch) (1776 bytes) - added by d.vinella on 2015-05-29 08:10:07Z - 
* [7916-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/352924bb86192b01cd6d4997ed8d86a2adcc7ef1.patch) (1658 bytes) - added by d.vinella on 2015-05-29 10:28:54Z - 
* [7916-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/63962b8f102c59c647697d1059a48df12ce214d7.patch) (3700 bytes) - added by d.vinella on 2015-05-31 10:48:52Z - 
* [7916-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/723aae8f8b43b0788ceed0c9eb4039bbf487ba41.patch) (2460 bytes) - added by d.vinella on 2015-06-01 15:39:12Z - 
* [changes-on-top-of-7916-5.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bd1/bd11e20bbb271cab345d27ca72bbff6b7a874e01/d90b42c6ad84b1d25235a5b25728ae173024b17a.diff) (3447 bytes) - added by adiroiban on 2015-06-08 08:26:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7916 7916
type__defect defect
reporter__d_vinella d.vinella
priority__normal normal
milestone__ 
branch__branches_beginLoggingTo_7916_2 branches/beginLoggingTo-7916-2
branch_author__hawkowl__adiroiban hawkowl, adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1432728305472527 1432728305472527
changetime__1437294525323795 1437294525323795
version__None None
owner__hawkowl hawkowl
cc__d.vinella
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to d.vinella|
|-|-|

Hi d.vinella!

Thanks for spotting this problem and submitting a patch.

However, all changes to Twisted require test coverage of some kind, to ensure the change has the desired effect.  In this case, clearly this broken format string was not tested previously ;).  Can you add a test - it can be very simple and short - just to verify that the change has had the intended effect, and will not regress in the future?

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|d.vinella removed owner|
|-|-|

Thank you Glyph for the review, I've attached the patch with a simple test, please review it!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set owner to d.vinella|
|-|-|

Hi,

Thanks for you changes!

----

Why is MORE_THAN_ONCE_WARNING public and why is it not documented?

This is not related to this ticket but I think that we should update that code.

-------

I feel that the new test is very fragile as it will always need to be kept in sync with test_beginLoggingToTwice 

test_beginLoggingToTwice has a more solid arrange phase... but it has a fragile assertion phase.

Maybe the test_beginLoggingToTwice test can be updated to call formatEvent(events2[0]) or compareEvents updated to also call formatEvent, or instead of using events1.append we define a log observer which will call formatEvent.

I found the need for calling formatEvent in other tests and I am in favor or creating a dedicated log observer to help with testing.

In case MORE_THAN_ONCE_WARNING really needs to be public, then yes we need a separate test for it to make sure its interface will never change.

I don't think that this is ready to merge.

Please review my comment.

Thank again!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|d.vinella removed owner|
|-|-|

Hi adiroiban, 
thanks for your review!

I'm not quite sure how the documentation for a module level string should look 
like; just a comment above it?
To keep this "string formatting" testing decoupled from stuff around it (logger
stuff in this case) maybe we could just implement formatting tests that just 
check formatWithCall does not throw an exception (like in attached draft 
7916-3.patch)? Moreover there are other strings like this one in other modules 
(e.g. logger._observer.OBSERVER_DISABLED): should all of them receive a similar
test?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Thanks for doing the review, Adi.

Replying to [adiroiban](#note_4):

> Why is MORE_THAN_ONCE_WARNING public

It's not public; it's in a private module, and it is never re-exported.

Also, that isn't changed in this patch :).

> and why is it not documented?

because it's not public, and it's an internal implementation detail of a private module :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [d.vinella](#note_5):
> Hi adiroiban, 
> thanks for your review!
> 
> I'm not quite sure how the documentation for a module level string should look 
> like; just a comment above it?

I don't think you should be documenting it here, but just for future reference, the way you document module-global variables is a `@var` in the module docstring.  So, in this case, 

```
#!py

"""
This module includes process-global state associated with the logging system,
and implementation of logic for managing that global state.

@var MORE_THAN_ONCE_WARNING: This probably shouldn't be documented, but here's
    where the documentation would go if it were.
@type MORE_THAN_ONCE_WARNING: native L{str}
"""
```

> To keep this "string formatting" testing decoupled from stuff around it (logger
> stuff in this case) maybe we could just implement formatting tests that just 
> check formatWithCall does not throw an exception (like in attached draft 
> 7916-3.patch)?

Given that this variable is more or less an implementation detail, I'd rather the test have that coupling, and test the log message emission itself in context.  Since you can easily set up in-memory log observers, or just have a local observer on your one `LogBeginner` instance's `_log` attribute, this test can pull in enough dependencies to be realistic without pulling in enough to start having ugly side-effects outside the test or requiring complicated mocks.

> Moreover there are other strings like this one in other modules 
> (e.g. logger._observer.OBSERVER_DISABLED): should all of them receive a similar
> test?

Perhaps later (file a separate ticket?) but not in this patch.  Always avoid scope creep - fix things like this one at a time; once you have a ticket that grows to be "let's fix all of X" it gets very hard to review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to d.vinella|
|-|-|

Thanks for the explanations, including the epydoc part. My bad for not reading the whole code.

As commented by Glyph, since the variable is not public there is no need to write public documentation and I think that there is no need for a separate test.

The patch also needs a newsfile.

Please review the latest comments and submit a new patch.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|d.vinella removed owner|
|-|-|

Thank you Adi and Glyph for the explainatins and reviews.
I attached a new patch that checks the formatting in the same test_beginLoggingToTwice test using a StringIO based textFileLogObserver.

Please review it!
Thanks

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Thanks! Looks better.

Do we need to context manager for StringIO?

This is not a real file with a global state so we can ignore its cleanup in tests.

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|d.vinella commented|
|-|-|

Hi Adi,

I just adhered to the style found in t.logger.test.test_file for the StringIO context manager. Now I've attached the 7916-5.patch without it.
Please review it!

thanks

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban commented|
|-|-|

(In [45009]) Branching to beginLoggingTo-7916.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

Thanks for the changes! They look good.

I have applied your patch and sent it to the builders.

I find it hard to understand the new assertion. I don't know what it tries to check and how it does it.

Also, I don't like that each logger uses 2 observers. I was expecting to use a single obeserver for each logger and that the observer is capable of all operations required to support the test.

I have attached a small patch on top of your patch/current branch.
Note that my patch is a quick hack and with it other tests are failing.

I am leaving this on the review queue so that other developers can check it.

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|d.vinella commented|
|-|-|

Hi Adi,

thanks for your review.
The assertions I wrote are meant to test the smallest possible substring of the
warning string object of this ticket to ensure:
 - a future simple rephrasing of it would not break the test
 - to catch a format error (subject of this ticket)
I felt that the previous
self.assertNotIn("Unable to format event", result)
was quite fragile and subject to an eventual rephrasing of the 
"Unable to format event" string itself.

I had a look at your changes-on-top-of-7916-5.diff, but I think that the 
comparison of the formatEvent outputs can miss the formatting problem. Using the
original malformed warning string the test fails because the reprs (printed by 
formatUnformattableEvent) of the events dicts are different. But they are 
different because of the absence of 'log_namespace', 'log_source', 'log_logger', 
'log_time' keys in the "warning" test dict. If one adds them to it with expected 
values (hacking something for log_time), the test would pass, even with the 
unformattable string.
Please correct me if I'm wrong.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

You are right. My patch is broken. I only wanted to get the direction and start a conversation.

This is one reason why I left the ticket in the review queue.

The patch should use a custom formatEvent which will just raise an exception, rather than handling it with formatUnformattableEvent

I think that we should wait for another developer and see how we should merge it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

(In [45247]) Branching to beginLoggingTo-7916-2.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl set owner to @hawkowl|
|-|-|

This patch seems small, sweet, and fixes up the issue.

The newsfile was a bit confusing, so I changed it a bit to explain what it was fixing for the user.

I'm waiting for the buildbots to go green (which I expect them to), and then I'll merge.

**Comment by twisted-trac** [comment]
|<img alt="d.vinella's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|d.vinella commented|
|-|-|

Thank you for the review hawkowl

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45274]) Merge beginLoggingTo-7916-2: missing braces and typo in duplicated beginLoggingTo warning string

Author: d.vinella
Reviewer: glyph, adiroiban, hawkowl
Fixes: [#7916](https://github.com/twisted/twisted/issues/7916)
